core/cli/cli.hpp

```diff
@@ -0,0 +1,139 @@
+      return _;                                             \
+    }                                                       \
+    auto *operator->() const {                              \
+      return &_;                                            \
+    }                                                       \
+    auto *operator->() {                                    \
+      return &_;                                            \
+    }                                                       \
+  }
+
+#define CLI_OPTIONAL(NAME, DESCRIPTION, TYPE)                              \
+  struct {                                                                 \
+    boost::optional<TYPE> _;                                               \
+    void operator()(Opts &opts) {                                          \
+      opts.add_options()(NAME, po::value(&_), DESCRIPTION);                \
+    }                                                                      \
+    operator bool() const {                                                \
+      return _.operator bool();                                            \
+    }                                                                      \
+    void check() const {                                                   \
+      if (!_) {                                                            \
+        throw ::fc::cli::CliError{"--{} argument is required but missing", \
+                                  NAME};                                   \
+      }                                                                    \
+    }                                                                      \
+    auto &operator*() const {                                              \
+      check();                                                             \
+      return *_;                                                           \
+    }                                                                      \
+    auto &operator*() {                                                    \
+      check();                                                             \
+      return *_;                                                           \
+    }                                                                      \
+    auto *operator->() const {                                             \
+      return &**this;                                                      \
+    }                                                                      \
+    auto *operator->() {                                                   \
+      return &**this;                                                      \
+    }                                                                      \
+  }
+
+#define CLI_OPTS() ::fc::cli::Opts opts()
+#define CLI_RUN()                  \
+  static ::fc::cli::RunResult run( \
+      ::fc::cli::ArgsMap &argm, Args &args, ::fc::cli::Argv &&argv)
```

Comment:
fields of ArgsMap should be changeable to avoid copying in case of cli fields modification is needed (f e create non constant reference via const_cast) const_cast will secure values during processing but will open it for modifying ib CLI_RUN()

Reply:
they are changeable now.